### PR TITLE
SNES: try to tell apart register lengths

### PR DESCRIPTION
**Pull request by usrshare**
This (admittedly very hacky) pull request tries to make sure the disassembler and analyzer handles X and M flags, which determine whether the X/Y registers and the accumulator are 8 or 16 bits long, properly.

Whenever either the disassembler or the analyzer sees a "rep" or "sep" instruction that would have modified the X and M flags on a real 65816 CPU, it sets an internal flag that makes sure the following instructions are processed with proper register length.

This is especially important when handling instructions that load data with immediate addressing (the value is specified in bytes following the opcode) or when comparing registers with an immediate value, as length of these instructions is influenced by the flags' values.

### CODE REVIEW

**Comment by Maijin** [comment]
(unrelated, what about a SNES rom file format https://github.com/radare/radare2/wiki/Implementing-a-new-format ) ?

File: libr/asm/p/asm_snes.c [hunk 1]
```diff
@@ -5,11 +5,33 @@
 #include <r_asm.h>
 #include <r_lib.h>
 #include "../arch/snes/snesdis.c"
+#include "asm_snes.h"
+
+struct snes_asm_flags* snesflags = NULL;
```
Comment (radare) [inline libr/asm/p/asm_snes.c]: static? globals are bad

File: libr/asm/p/asm_snes.c [hunk 2]
```diff
@@ -5,11 +5,33 @@
...
+
+bool snes_asm_init (void* user) {
+	if (!snesflags) snesflags = malloc(sizeof( struct snes_asm_flags ));
```
Comment (radare) [inline libr/asm/p/asm_snes.c]: use r2-indent (braces on all if blocks, even the oneliner ones). add spaces *before* the (, not after

File: libr/asm/p/asm_snes.c [hunk 3]
```diff
@@ -5,11 +5,33 @@
...
+	memset(snesflags,0,sizeof (struct snes_asm_flags));
```
Comment (radare) [inline libr/asm/p/asm_snes.c]: add spaces after every , or just use r2-indent

File: libr/asm/p/asm_snes.c [hunk 4]
```diff
@@ -5,11 +5,33 @@
...
+	return 0;
+}
 
 static int disassemble(RAsm *a, RAsmOp *op, const ut8 *buf, int len) {
-	int dlen = snesDisass (a->bits, a->pc, op, buf, len);
+	int dlen = snesDisass (snesflags->M, snesflags->X, a->pc, op, buf, len);
 	if (dlen<0) dlen=0;
 	op->size = dlen;
+	if (buf[0] == 0xc2) { //REP
+		if ( buf[1] & 0x10 ) snesflags->X = 0;
```
Comment (radare) [inline libr/asm/p/asm_snes.c]: remove spaces after ( and before )

File: libr/asm/p/asm_snes.c [hunk 5]
```diff
@@ -5,11 +5,33 @@
...
+bool snes_asm_fini (void* user) {
+	free(snesflags);
+	snesflags = NULL;
+	return 0;
+}
 
 static int disassemble(RAsm *a, RAsmOp *op, const ut8 *buf, int len) {
-	int dlen = snesDisass (a->bits, a->pc, op, buf, len);
+	int dlen = snesDisass (snesflags->M, snesflags->X, a->pc, op, buf, len);
 	if (dlen<0) dlen=0;
```
Comment (radare) [inline libr/asm/p/asm_snes.c]: use r2-indent

**Comment by radare** [comment]
i'll merge and r2-indent after that to avoid messing with the change too much.

ill also fix the broken test which is harmless.

thanks!

**Comment by usrshare** [comment]
I apologize for not following the indentation rules.

In a related note, is it possible to modify the interface for _init / _fini functions, so that, they are called with one of the arguments being the pointer to the respective structure being initialized or destroyed?

For example, in my case, I would love to use the _set_user_ptr functions to store a pointer to a structure that would have held the current flag values instead of using a global variable, but, since to set a user pointer requires having a pointer to the RAsm / RAnal structure, which at this point is unavailable, this is currently impossible.

Basically, I'd like to see if the format for these functions can be changed from

    bool (*init)(void *user);

to

    bool (*init)(RAsm* a, void* user);

**Comment by radare** [comment]
yes this change makes sense. already did in other modules like rbin or so.. send a pr with this change please
